pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -705,17 +705,59 @@ func proxyNeedsPush(con *XdsConnection, targetNamespaces map[string]struct{}) bo
 	return false
 }
 
+// updateProxyServiceInstances determines whether the proxy associated with the given service instance
+// needs to refresh its local inbound listeners, and if so, triggers a push to that proxy.
+func (s *DiscoveryServer) updateProxyServiceInstances(instance *model.ServiceInstance) error {
+	instanceIP := instance.Endpoint.Address
+
+	// Check whether there's an existing connection from that IP address
+	adsClientsMutex.RLock()
+	con, ok := s.connectionsByIP[instanceIP]
+	adsClientsMutex.RUnlock()
+	if !ok {
+		return nil
+	}
+
+	proxy := con.modelNode
+	instances, err := s.Env.GetProxyServiceInstances(proxy)
+	if err != nil {
+		adsLog.Errorf("Error getting proxy %s service instances: %v", proxy.ID, err)
+		return err
+	}
+
+	if reflect.DeepEqual(instances, proxy.ServiceInstances) {
+		// no changes detected - nothing to update
+		return nil
+	}
+
+	// Update proxy node's service instances
+	err = proxy.SetServiceInstances(s.Env)
```

Comment:
in SetServiceInstances we call GetProxyServiceInstances, but we just called this before (as `instances`). Should we just do `proxy.ServiceInstances = instances`?

Reply:
Also, I suspect this will cause a race condition, because ServiceInstances is not protected by a mutex -- none of the fields in model.Proxy are because we always call them from a single goroutine, but this would break that

Reply:
Done regarding `proxy.ServiceInstances = instances`.
And yes, good observation about the race condition. That single goroutine you mentioned is the one that runs `StreamAggregatedResources()`, right? If so, I'll work on adding it another input channel that will be populated by the above handler func.